[WebAuthn] Populate excludeCredentials in modern AS flow

pascoej opened the pull request:
#### 3e73d2ab80517c18e34b74ea0c985739698f3baa
<pre>
[WebAuthn] Populate excludeCredentials in modern AS flow
<a href="https://bugs.webkit.org/show_bug.cgi?id=270856">https://bugs.webkit.org/show_bug.cgi?id=270856</a>
<a href="https://rdar.apple.com/124405037">rdar://124405037</a>

Reviewed by Chris Dumez.

This patch starts passing along the excluded credentials from the request along the modern
AS flow.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):

Canonical link: <a href="https://commits.webkit.org/276006@main">https://commits.webkit.org/276006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fba3232e3bc4b99a4abbe67d1de41edad3f71cec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43504 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46139 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39633 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35972 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19585 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16946 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1560 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38848 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47683 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/18534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15166 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42755 "Failure limit exceed. At least found 1 new test failure: imported/w3c/web-platform-tests/webrtc-extensions/transfer-datachannel-service-worker.https.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19957 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41426 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9678 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/20136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19587 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -198,6 +198,55 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
+    ExceptionCode exceptionCode = (ExceptionCode)nsErrorCode;
+
+    switch (exceptionCode) {
+    case ExceptionCode::IndexSizeError: FALLTHROUGH;
```
I doubt you need all these `FALLTHROUGH;` calls since those cases have no body.

pvollan committed:
Use system import directory when compiling sandboxes
https://bugs.webkit.org/show_bug.cgi?id=270868
rdar://124373752

Reviewed by Chris Dumez.

Use system import directory when compiling sandboxes on macOS. This fixes a build issue seen on some
host configurations.

This patch was written by Christian Schmidt and Per Arne Vollan.

* Source/WebKit/Scripts/compile-sandbox.sh:

Canonical link: https://commits.webkit.org/275996@main

jyavenard committed:
<audio> and <video> loadeddata events not fired on page load
https://bugs.webkit.org/show_bug.cgi?id=270837
rdar://124079735

Reviewed by Eric Carlson.

The HTMLMediaElement would call MediaPlayer::prepareToPlay() on the first MediaPlayerPrivate
created which may not be the one we end up using.
The code assumed that as soon as a MediaPlayer was created, we could call prepareToPlay on
it which in the case of the MediaPlayerPrivateAVFobjC would start loading the content.
Since we enabled the WebM player, the assumption no longer applied, multiple players could
be used until we find one that can play the content.
If the GPU process was enabled, the behaviour was racy as the GPUP's MediaPlayer
may not have been created yet.

The site would set a HTMLMediaElement's source to a mp3 file, without using an explicit
extension nor having the server provide the mime-type. As such, we have to try in
succession all MediaPlayerPrivate until we can find one that can load the content.
We cache the call to prepareToPlay() and re-issue it on all new MediaPlayerPrivate once
created.

Added test.

* LayoutTests/http/tests/media/audio-load-loadeddata-expected.txt: Added.
* LayoutTests/http/tests/media/audio-load-loadeddata.html: Added.
* LayoutTests/platform/ios/TestExpectations: All media tests are disabled on iOS, force this one to run.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::loadWithNextMediaEngine):
(WebCore::MediaPlayer::prepareToPlay):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::prepareForPlayback):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::prepareForPlayback):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::prepareForPlayback):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:

Canonical link: https://commits.webkit.org/275997@main

minorninth committed:
AX: VoiceOver doesn't read characters when user presses left / right arrows in Monaco code editor
https://bugs.webkit.org/show_bug.cgi?id=270616
rdar://123984168

Reviewed by Andres Gonzalez.

When the character or selection extent moves by just one visible position, infer
that it was a character granularity move, rather than a discontiguous selection.

* LayoutTests/accessibility/mac/custom-text-editor-expected.txt: Added.
* LayoutTests/accessibility/mac/custom-text-editor.html: Added.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::inferDirectionFromIntent):
(WebCore::AXObjectCache::postTextStateChangePlatformNotification):

Canonical link: https://commits.webkit.org/275998@main

kiaraarose committed:
Cannot manage extension storage with WebKit extensions in Safari
https://bugs.webkit.org/show_bug.cgi?id=270869
rdar://124037961

Reviewed by Timothy Hatcher.

Update methods for fetching and removing data records. Since it's not guarenteed the context
will be loaded, don't refer to the controller's map of extension context. Instead, do a direct
comparision with the uniqueIdentifier from the context passed to determine if a match is found.

Also, update _storageDirectoryPath so that a custom storage path can be set.

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
(-[_WKWebExtensionSQLiteStore _deleteDatabaseFileAtURL:reopenDatabase:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(-[_WKWebExtensionControllerConfiguration _setStorageDirectoryPath:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::readDisplayNameFromState):
(WebKit::WebExtensionContext::invalidateStorage):
Make this public so it can be used after sql dbs are closed when data records are removed.
(WebKit::WebExtensionContext::readDisplayNameAndLastBaseURLFromState): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::sqliteStore):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfigurationCocoa.mm:
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath):
Convert identifier back to uppercase after toString(). This was causing a bug where
a new (lowercased) directory would be used for extension storage.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:

Canonical link: https://commits.webkit.org/275999@main

AndresGonzalezApple committed:
AX: AccessibilityNodeObject::determineAccessibilityRoleFromNode makes many unnecessary dynamicDowncasts.
https://bugs.webkit.org/show_bug.cgi?id=270854
<rdar://problem/124455559>

Reviewed by Tyler Wilcock.

This is a hot method in ITM that calls many times Node::hasTagName(const HTMLQualifiedName&), which in turn dynamicDowncasts the Node to an HTMLElement. This patch reduces the number of dynamicDowncasts significantly by downcasting to an HTMLElement upfront and calling HTMLElement::hasTagName directly instead. In a run of LayoutTests/accessibility, the number of downcasts drops by 1978248 or 13% of the original number.

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):

Canonical link: https://commits.webkit.org/276000@main

aprotyas committed:
Always use DDHighlightCreateWithRectsInVisibleRectWithStyleAndDirection in DDHighlight creation code
https://bugs.webkit.org/show_bug.cgi?id=270865
rdar://124465778

Reviewed by Tim Horton and Megan Gardner.

Currently, we call DDHighlightCreateWithRectsInVisibleRectWithStyleAndDirection
conditional on HAVE(DD_HIGHLIGHT_CREATE_WITH_SCALE). This is not
necessary because all calls to the DDHighlight creation method happen
in macOS-only codepaths, and the method in question is available on all
supported macOS configurations.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.mm:
* Source/WebCore/page/mac/ImageOverlayControllerMac.mm:
(WebCore::ImageOverlayController::updateDataDetectorHighlights):
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildPhoneNumberHighlights):
(WebCore::ServicesOverlayController::buildSelectionHighlight):

Canonical link: https://commits.webkit.org/276001@main

charliewolfe committed:
Null pointer dereference in `U2fAuthenticator::issueSignCommand`
https://bugs.webkit.org/show_bug.cgi?id=270873
rdar://124475802

Reviewed by Pascoe.

Crash logs indicate it's possible for `u2fCmd` to be null. Avoid crashing in this case.

* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::issueSignCommand):

Canonical link: https://commits.webkit.org/276002@main

xeenon committed:
REGRESSION (275981@main): [ macOS ] 7 tests in TestWebKitAPI.WKWebExtensionAPIPermissions are a constant failure/timeout.
https://webkit.org/b/270872
rdar://problem/124475393

Reviewed by Jeff Miller.

When requesting permissions via permissions.request(), we need to include optional permissions
in the check. That was actually happening by accident before, since !hasPermission() was returning
true for any unknown permission. With the change to needsPermission(), only RequestedExplicitly
or RequestedImplicitly results were counted. To fix this, a new option was added for use by
the permissions API to include optional permissions in the RequestedImplicitly category.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
(WebKit::WebExtensionContext::permissionsRequest): Pass the IncludeOptionalPermissions option.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::requestPermissionMatchPatterns): Pass through options to needsPermission().
(WebKit::WebExtensionContext::requestPermissionToAccessURLs): Ditto.
(WebKit::WebExtensionContext::requestPermissions): Ditto.
(WebKit::WebExtensionContext::permissionState): Added support for IncludeOptionalPermissions to check
the optional permissions from the extension's manifest.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::patternsMatchURL): Added const to URL to fix build issue.
(WebKit::WebExtensionMatchPattern::patternsMatchPattern): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::requestPermissionMatchPatterns): Added options parameter.
(WebKit::WebExtensionContext::requestPermissionToAccessURLs): Ditto.
(WebKit::WebExtensionContext::requestPermissions): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TEST(WKWebExtensionAPITabs, QueryWithAccessPrompt)): Grant tabs permission so the tab URL counts as a
requested URL and the prompt delegate will be called.

Canonical link: https://commits.webkit.org/276003@main

alanbaradlay committed:
[RenderTreeDump] Use linesBoundingBox for text, line break and inline box
https://bugs.webkit.org/show_bug.cgi?id=270731#c19

Unreviewed rebaseline after 275888@main.

* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/text/visualtests/text-inline-size-003-visual-expected.txt:
* LayoutTests/platform/ios/fast/block/basic/001-expected.txt:

Canonical link: https://commits.webkit.org/276004@main

achristensen07 committed:
Fix some assertions when using a canvas without a layer with webgpu
https://bugs.webkit.org/show_bug.cgi?id=270705
rdar://124213314

Reviewed by Mike Wyrzykowski.

The test in this PR hit several issues:
1. PresentationContextIOSurface::present asserted, which I changed to an early return.
2. I handle errors when creating objects by returning nullptr instead of an object that
   has no corresponding object in the GPU process, and propogate those errors to JS.
3. I added a WeakPtr check in the lambda in GPUCanvasContextCocoa::prepareForDisplay
4. In order to keep Windows building after this change, I needed to change
   FontPlatformData::Attributes to FontPlatformDataAttributes in WebKit.

* LayoutTests/fast/webgpu/use-canvas-without-layer-expected.txt: Added.
* LayoutTests/fast/webgpu/use-canvas-without-layer.html: Added.
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessages):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::~RemoteTextureProxy):
(WebKit::WebGPU::RemoteTextureProxy::createView):
(WebKit::WebGPU::RemoteTextureProxy::destroy):
(WebKit::WebGPU::RemoteTextureProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:

Canonical link: https://commits.webkit.org/276005@main

pascoej committed:
[WebAuthn] Populate excludeCredentials in modern AS flow
https://bugs.webkit.org/show_bug.cgi?id=270856
rdar://124405037

Reviewed by Chris Dumez.

This patch starts passing along the excluded credentials from the request along the modern
AS flow.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegisteration):

Canonical link: https://commits.webkit.org/276006@main
